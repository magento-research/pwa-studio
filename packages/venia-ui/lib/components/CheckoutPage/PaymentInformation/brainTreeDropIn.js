/**
 * @fileoverview This component uses the BrainTree Web Drop In to hook into Web
 * Payments and the Payment Request API to submit payments via BrainTree.
 *
 * @see
 *   https://github.com/braintree/braintree-web-drop-in
 *   https://braintree.github.io/braintree-web-drop-in/docs/current/index.html
 *   https://developers.braintreepayments.com/guides/drop-in/overview/javascript/v3
 *   https://developers.braintreepayments.com/guides/payment-method-nonce.
 *   https://braintree.github.io/braintree-web-drop-in/docs/current/Dropin.html#requestPaymentMethod.
 */

import React, { useEffect, useState, useCallback } from 'react';
import { bool, func, shape, string } from 'prop-types';

import defaultClasses from './braintreeDropin.css';
import { mergeClasses } from '../../../classify';

const authorization = process.env.CHECKOUT_BRAINTREE_TOKEN;

/**
 * This BraintreeDropin component has two purposes which lend to its
 * implementation:
 *
 * 1) Mount and asynchronously create the dropin via the braintree api.
 * 2) On submission (triggered by a parent), request the payment nonce.
 */
const BraintreeDropin = props => {
    const {
        onError,
        onReady,
        onSuccess,
        shouldRequestPaymentNonce,
        containerId,
        shouldTeardownDropin,
        resetShouldTeardownDropin
    } = props;
    const classes = mergeClasses(defaultClasses, props.classes);
    const [isError, setIsError] = useState(false);
    const [dropinInstance, setDropinInstance] = useState();

    const createDropinInstance = useCallback(
        async didClose => {
            try {
                const {
                    default: dropIn
                } = await import('braintree-web-drop-in');
                const dropinInstance = await dropIn.create({
                    authorization,
                    container: `#${containerId}`,
                    card: {
                        cardholderName: {
                            required: true
                        },
                        overrides: {
                            fields: {
                                number: {
                                    maskInput: {
                                        // Only show last four digits on blur.
                                        showLastFour: true
                                    }
                                }
                            }
                        }
                    }
                });

                if (didClose) {
                    // If we get here but the form is closed we should teardown
                    // instead of trying to use the instance.
                    dropinInstance.teardown();
                } else {
                    setDropinInstance(dropinInstance);
                    onReady(true);
                }
            } catch (err) {
                console.error(
                    `Unable to initialize Credit Card form (Braintree). \n${err}`
                );

                if (!didClose) {
                    // Error state used to render error view. If we got an error
                    // but we are closing the form we shouldn't try to update
                    // state.
                    setIsError(true);
                }
            }
        },
        [onReady, containerId]
    );

    useEffect(() => {
        let didClose = false;

        // Initialize the dropin with a reference to the container mounted in
        // this component. We do this via onmount because we have to be sure
        // the container id has mounted per braintree's API.
        createDropinInstance(didClose);

        // If we close the form quickly after opening we may end up in a
        // semi-mounted state so we need a local variable to make sure we
        // behave correctly.
        return () => {
            didClose = true;
        };
    }, [createDropinInstance]);

    useEffect(() => {
        async function requestPaymentNonce() {
            try {
                const paymentNonce = await dropinInstance.requestPaymentMethod();
                onSuccess(paymentNonce);
            } catch (e) {
                // An error occurred. BrainTree will update the UI with error
                // messaging, but we should signal that there was an error.
                console.error(`Invalid Payment Details. \n${e}`);
                onError();
            }
        }

        if (shouldRequestPaymentNonce) {
            requestPaymentNonce();
        }
    }, [dropinInstance, onError, onSuccess, shouldRequestPaymentNonce]);

    useEffect(() => {
        if (shouldTeardownDropin) {
            dropinInstance
                .teardown()
                .then(() => {
                    resetShouldTeardownDropin();
                    createDropinInstance();
                })
                .catch(err => {
                    console.error('Teardown failed', err);
                });
        }
    }, [
        shouldTeardownDropin,
        dropinInstance,
        resetShouldTeardownDropin,
        createDropinInstance
    ]);

    if (isError) {
        return (
            <span className={classes.error}>
                There was an error loading payment options. Please try again
                later.
            </span>
        );
    }

    return (
        <div className={classes.root}>
            <div id={containerId} />
        </div>
    );
};

export default BraintreeDropin;

BraintreeDropin.propTypes = {
    classes: shape({
        root: string,
        error: string
    }),
    containerId: string.isRequired,
    onError: func.isRequired,
    onReady: func.isRequired,
    onSuccess: func.isRequired,
    shouldRequestPaymentNonce: bool.isRequired
};
