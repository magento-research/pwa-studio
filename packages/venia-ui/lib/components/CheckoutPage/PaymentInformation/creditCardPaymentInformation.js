import React from 'react';

import BrainTreeDropin from '../../Checkout/braintreeDropin';

const CreditCardPaymentInformation = props => {
    return (
        <BrainTreeDropin
            containerID="checkout_page_payment_information"
            onError={console.error}
            onReady={console.log}
            onSuccess={console.warn}
            shouldRequestPaymentNonce={false}
        />
    );
};

export default CreditCardPaymentInformation;
