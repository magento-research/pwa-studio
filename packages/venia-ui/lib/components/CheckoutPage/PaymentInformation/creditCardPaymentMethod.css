.root {
}

.dropin_root {
    visibility: visible;
    opacity: 1;
    transition-delay: 64ms;
    transition-duration: 384ms;
    transition-property: opacity, visbility;
    transition-timing-function: var(--venia-anim-standard);
}

.dropin_hidden {
    visibility: hidden;
    height: 0;
    opacity: 0;
}

.billing_address_fields_root {
    display: grid;
    gap: 0.5rem;
    padding: 1rem 0rem;
}

.errors_container {
    display: grid;
    gap: 0.5rem;
}

.error {
    color: red;
}

.first_name,
.last_name,
.city,
.state,
.postal_code,
.phone_number {
    grid-column-end: span 1;
}

.country,
.street1,
.street2,
.address_check {
    grid-column-end: span 2;
}

/*
 * Mobile-specific styles.
 */

@media (max-width: 960px) {
    .first_name,
    .last_name,
    .city,
    .state,
    .postal_code,
    .phone_number,
    .street1,
    .street2,
    .address_check {
        grid-column-end: span 2;
    }
}
