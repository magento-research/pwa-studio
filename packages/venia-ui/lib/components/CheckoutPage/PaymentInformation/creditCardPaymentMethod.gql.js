import gql from 'graphql-tag';

// We disable linting for local fields because there is no way to add them to
// the fetched schema.
// https://github.com/apollographql/eslint-plugin-graphql/issues/99
/* eslint-disable graphql/template-strings */

export const GET_ALL_COUNTRIES = gql`
    query getAllCountries {
        countries {
            available_regions {
                code
                id
                name
            }
            id
        }
    }
`;

export const GET_IS_BILLING_ADDRESS_SAME = gql`
    query getIsBillingAddressSame($cartId: String) {
        cart(cart_id: $cartId) {
            id
            isBillingAddressSame @client
        }
    }
`;

export const GET_BILLING_ADDRESS = gql`
    query getBillingAddress($cartId: String) {
        cart(cart_id: $cartId) {
            id
            billingAddress @client {
                firstName
                lastName
                street1
                street2
                city
                state
                postalCode
                phoneNumber
            }
        }
    }
`;

export const GET_PAYMENT_NONCE = gql`
    query getPaymentNonce($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            paymentNonce @client
        }
    }
`;

/* eslint-enable graphql/template-strings */

export default {
    queries: {
        getAllCountriesQuery: GET_ALL_COUNTRIES,
        getBillingAddressQuery: GET_BILLING_ADDRESS,
        getIsBillingAddressSameQuery: GET_IS_BILLING_ADDRESS_SAME,
        getPaymentNonceQuery: GET_PAYMENT_NONCE
    },
    mutations: {}
};
