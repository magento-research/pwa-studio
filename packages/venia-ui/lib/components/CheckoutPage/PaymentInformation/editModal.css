.root {
    background-color: white;
    height: auto;
    left: calc(50% - 370px);
    max-height: 85vh;
    max-width: 740px;
    opacity: 0;
    overflow: hidden;
    position: fixed;
    top: 5vh;
    transform: scale(1.15);
    transition-duration: 192ms;
    transition-timing-function: var(--venia-anim-out);
    transition-property: opacity, transform, visibility;
    visibility: hidden;
    width: 100%;
    z-index: 3;
}

.root_open {
    composes: root;
    box-shadow: 1px 0 rgb(var(--venia-border));
    opacity: 1;
    transform: scale(1);
    transition-duration: 224ms;
    transition-timing-function: var(--venia-anim-in);
    visibility: visible;
}

.body {
    max-height: calc(85vh - 3.5rem);
    overflow: auto;
    padding: 0.5rem 2rem;
}

.header {
    border-bottom: 1px solid rgb(var(--venia-border));
    display: flex;
    justify-content: space-between;
    padding: 0.875rem 2rem;
}

.headerText {
    align-self: center;
    color: rgb(var(--venia-text-alt));
}

@media (max-width: 960px) {
    .root {
        bottom: 0;
        height: 100%;
        left: unset;
        max-height: unset;
        max-width: 360px;
        right: 0;
        top: 0;
        transform: translate3d(100%, 0, 0);
    }

    .root_open {
        composes: root;
        transform: translate3d(0, 0, 0);
    }

    .body {
        max-height: calc(100vh - 3.5rem);
        padding: 0.5rem 1rem;
    }

    .header {
        padding: 0.875rem;
    }
}
