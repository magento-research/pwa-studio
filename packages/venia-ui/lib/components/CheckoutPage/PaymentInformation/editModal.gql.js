import gql from 'graphql-tag';


export const GET_SELECTED_PAYMENT_METHOD = gql`
    query getSelectedPaymentMethod($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            selectedPaymentMethod: selected_payment_method {
                code
            }
        }
    }
`;

/* eslint-enable graphql/template-strings */

export default {
    queries: {
        getSelectedPaymentMethodQuery: GET_SELECTED_PAYMENT_METHOD
    },
    mutations: {}
};
