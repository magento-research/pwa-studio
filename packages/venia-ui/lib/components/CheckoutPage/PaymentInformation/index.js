import React from 'react';
import { Form } from 'informed';

import PaymentInformation from './paymentInformation';

const PaymentInformationWrapper = props => {
    return (
        <Form>
            <PaymentInformation {...props} />
        </Form>
    );
};

export default PaymentInformationWrapper;
