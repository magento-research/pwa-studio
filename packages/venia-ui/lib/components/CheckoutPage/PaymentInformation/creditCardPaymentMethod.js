import React from 'react';

import { useCreditCard } from '@magento/peregrine/lib/talons/CheckoutPage/useCreditCard';

import BrainTreeDropin from './brainTreeDropIn';

import defaultClasses from './creditCardPaymentMethod.css';

const CreditCardPaymentInformation = props => {
    const {
        shouldRequestPaymentNonce,
        isHidden,
        doneEditing,
        setDoneEditing
    } = props;
    const {
        onPaymentError,
        onPaymentSuccess,
        onPaymentReady,
        paymentNonce
    } = useCreditCard({ setDoneEditing });

    const creditCardComponent = isHidden ? null : doneEditing ? (
        <div className={defaultClasses.summary}>
            <span>Credit Card</span>
            <span>{`${paymentNonce.details.cardType} ending in ${
                paymentNonce.details.lastFour
            }`}</span>
        </div>
    ) : (
        <BrainTreeDropin
            onError={onPaymentError}
            onReady={onPaymentReady}
            onSuccess={onPaymentSuccess}
            shouldRequestPaymentNonce={shouldRequestPaymentNonce}
        />
    );

    return <div className={defaultClasses.root}>{creditCardComponent}</div>;
};

export default CreditCardPaymentInformation;
