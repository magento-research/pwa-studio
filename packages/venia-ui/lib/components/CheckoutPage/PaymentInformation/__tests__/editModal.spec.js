import React from 'react';
import ReactDOM from 'react-dom';
import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';
import { useEditModal } from '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useEditModal';

import Button from '../../../Button';
import CreditCardPaymentInformation from '../creditCardPaymentMethod';
import EditModal from '../editModal';

jest.mock('../../../../classify');

jest.mock(
    '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useEditModal',
    () => ({
        useEditModal: jest.fn().mockReturnValue({
            selectedPaymentMethod: 'braintree',
            isLoading: false,
            updateButtonClicked: false,
            handleClose: () => {},
            handleUpdate: () => {},
            handlePaymentSuccess: () => {},
            handleDropinReady: () => {}
        })
    })
);

jest.mock('../creditCardPaymentMethod', () => () => (
    <div>Credit Card Payment Method</div>
));

jest.mock('../../../Button', () => {
    return () => <div>Button Component</div>;
});

jest.mock('../../../Icon', () => {
    return () => <div>Icon Component</div>;
});

beforeAll(() => {
    /**
     * Mocking ReactDOM.createPortal because of incompatabilities
     * between ReactDOM and react-test-renderer.
     *
     * More info: https://github.com/facebook/react/issues/11565
     */
    ReactDOM.createPortal = jest.fn(element => {
        return element;
    });
});

afterAll(() => {
    ReactDOM.createPortal.mockClear();
});

test('Should return correct shape', () => {
    const tree = createTestInstance(<EditModal />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('Should render creditCardPaymentInformation component if selectedPaymentMethod is braintree', () => {
    useEditModal.mockReturnValueOnce({
        selectedPaymentMethod: 'braintree',
        isLoading: true,
        updateButtonClicked: false,
        handleClose: () => {},
        handleUpdate: () => {},
        handlePaymentSuccess: () => {},
        handleDropinReady: () => {}
    });

    const tree = createTestInstance(<EditModal />);

    expect(tree.root.findByType(CreditCardPaymentInformation)).not.toBeNull();
});

test('Should not render creditCardPaymentInformation component if selectedPaymentMethod is not braintree', () => {
    useEditModal.mockReturnValueOnce({
        selectedPaymentMethod: 'paypal',
        isLoading: true,
        updateButtonClicked: false,
        handleClose: () => {},
        handleUpdate: () => {},
        handlePaymentSuccess: () => {},
        handleDropinReady: () => {}
    });

    const tree = createTestInstance(<EditModal />);

    expect(() => {
        tree.root.findByType(CreditCardPaymentInformation);
    }).toThrow();
});

test('Should render Cancel and Update buttons only if isLoading is false', () => {
    useEditModal.mockReturnValueOnce({
        selectedPaymentMethod: 'braintree',
        isLoading: true,
        updateButtonClicked: false,
        handleClose: () => {},
        handleUpdate: () => {},
        handlePaymentSuccess: () => {},
        handleDropinReady: () => {}
    });

    const tree = createTestInstance(<EditModal />);

    expect(() => {
        tree.root.findByType(Button);
    }).toThrow();

    useEditModal.mockReturnValueOnce({
        selectedPaymentMethod: 'braintree',
        isLoading: false,
        updateButtonClicked: false,
        handleClose: () => {},
        handleUpdate: () => {},
        handlePaymentSuccess: () => {},
        handleDropinReady: () => {}
    });

    tree.update(<EditModal />);

    expect(tree.root.findAllByType(Button)).toHaveLength(2);
});

test('Actions buttons should be disabled if updateButtonClicked is true', () => {
    useEditModal.mockReturnValueOnce({
        selectedPaymentMethod: 'braintree',
        isLoading: false,
        updateButtonClicked: true,
        handleClose: () => {},
        handleUpdate: () => {},
        handlePaymentSuccess: () => {},
        handleDropinReady: () => {}
    });

    const tree = createTestInstance(<EditModal />);

    expect(
        tree.root.findAllByType(Button).map(({ props }) => props.disabled)
    ).toMatchObject([true, true]);
});
