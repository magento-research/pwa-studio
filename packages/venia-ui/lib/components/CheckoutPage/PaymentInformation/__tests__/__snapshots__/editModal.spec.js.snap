// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
<aside
  className="root"
>
  <form
    className="form"
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <button
      className="mask"
      disabled={false}
      onClick={[Function]}
      type="reset"
    />
    <div
      className="dialog"
    >
      <div
        className="header"
      >
        <span
          className="headerText"
        >
          Edit Payment Information
        </span>
        <button
          className="headerButton"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          <mock-ChecIconkbox
            src={[Function]}
          />
        </button>
      </div>
      <div
        className="body"
      >
        <div
          className="contents"
        >
          <div
            className="body"
          >
            <mock-CreditCard
              onDropinReady={[Function]}
              onPaymentSuccess={[Function]}
              shouldSubmit={false}
            />
          </div>
        </div>
        <div
          className="buttons"
        >
          <mock-Button
            classes={
              Object {
                "root_lowPriority": "cancelButton",
              }
            }
            disabled={false}
            onClick={[Function]}
            priority="low"
            type="reset"
          >
            Cancel
          </mock-Button>
          <mock-Button
            classes={
              Object {
                "root_highPriority": "confirmButton",
              }
            }
            disabled={false}
            priority="high"
            type="submit"
          >
            Update
          </mock-Button>
        </div>
      </div>
    </div>
  </form>
</aside>
`;
